### PR TITLE
docs: add CONTRIBUTING.md and MAINTAINERS.md

### PR DESCRIPTION
**Pull request by theoludwig**
As the project is gaining activity again, and new contributors arise, it might be the right moment to define what are the "rules" for contributing to this project, and how pull requests are reviewed and merged.

Ref: #364 

(please take note, that English is not my primary language, so feel free to correct me :smile:)

### CODE REVIEW

File: CONTRIBUTING.md
Comment (theoludwig) [inline CONTRIBUTING.md:36]: Feel free to tell me what do you think about these "values". The goal is to make it so that everyone can share their feedbacks, and have the time to do so, without however, blocking the project and the "innovation". We might increase or decrease these requirements.

File: CONTRIBUTING.md
Comment (theoludwig) [inline CONTRIBUTING.md:67]: It has already been discussed previously: https://github.com/editorconfig-checker/editorconfig-checker/pull/350#issuecomment-2203634216

We might consider anything that results in **editorconfig-checker** reporting more linting errors, a major version.

**Comment by theoludwig** [comment]
cc @klaernie

File: CONTRIBUTING.md
Comment (per1234) [inline CONTRIBUTING.md:46]: It is great to define such a policy, but this is off topic for a contributing guide. It should go somewhere else other than CONTRIBUTING.md.

File: CONTRIBUTING.md
Comment (per1234) [inline CONTRIBUTING.md:24]: I'm not a fan of requiring an issue before a PR. It greatly increases the effort required for a contributor to make a simple contribution.

File: CONTRIBUTING.md
Comment (per1234) [inline CONTRIBUTING.md:38]: This is off topic for a contributing guide because most contributors will not be merging their PR.

I'm also not a fan of having a blanket "always squash" policy for project maintainers. When a pull request contains non-atomic "fixup" commits, they should definitely be squashed before merging. However, a contributor may intentionally put multiple atomic commits into a pull request and in this case squashing the commits is harmful.

**Review by per1234 — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (ccoVeille) [reply CONTRIBUTING.md:24]: I agree. If there is a need for an issue, we can ask the contributor to open one.

I often make PRs for fixing typos, requiring to open a PR for such PR would be counterproductive

File: CONTRIBUTING.md
Comment (ccoVeille) [reply CONTRIBUTING.md:38]: I agree. Talking about the process of merging is a bit out of contributions guide.

But at same time, having a clear policy is good for maintainers.

But that said, I also disagree about the fact every PR must be squashed. Sometimes separate commits are interesting.

So maybe the bguide could talk about the fact, that most PRs will be squashed.

File: CONTRIBUTING.md
Comment (ccoVeille) [inline CONTRIBUTING.md:44]: I'm fine with this, but then you should add something to CI to check all commits respect the policy.

It also requires to configure dependabot to use this pattern when opening PRs

File: CONTRIBUTING.md
Comment (theoludwig) [reply CONTRIBUTING.md:24]: Yes, not every PR needs to have first an issue open, I agree. I think the wording is misleading here, it was more a recommendation, that opening an issue first allows discussions and check if your goals match the goal of everyone else, before implementing something not wanted and "wasting time".

File: CONTRIBUTING.md
Comment (theoludwig) [reply CONTRIBUTING.md:46]: Okay, I'm fine with that, we can separate maintainers related and contributors related.

The main goal of this PR was to define maintainers policy, how could this file be named?

File: CONTRIBUTING.md
Comment (theoludwig) [reply CONTRIBUTING.md:38]: With the current settings on GitHub of this repo, PRs can only be squashed, it is not possible to merge them with multiple commits (however, this can of course change).

Also squash-and-merge doesn't erase history, as it will still be visible in the PR, and as said promotes clear separation of changes into multiple PRs, and therefore makes it easier to review.

File: CONTRIBUTING.md
Comment (klaernie) [reply CONTRIBUTING.md:46]: I think the SemVer policy belongs into the README - after all maintainers need to adhere to it, but users need to know it, no matter if they contribute or merely consume

File: CONTRIBUTING.md
Comment (klaernie) [reply CONTRIBUTING.md:24]: Maybe a wording like "please discuss bigger changes first" could make it clearer, that we encourage quick and simple PRs.

File: CONTRIBUTING.md
Comment (theoludwig) [reply CONTRIBUTING.md:44]: There is already `wagoid/commitlint-github-action` to check that commits respect the policy in CI.
And dependabot is already configured, to respect this commit policy. :smile:

It was however not documented.

Comment (klaernie) [reply CONTRIBUTING.md:44]: CI does that already - I learned that the hard way ;)

File: CONTRIBUTING.md
Comment (klaernie) [reply CONTRIBUTING.md:36]: I wonder if it would be better to specify the "open for 72h" as "48h of not requesting changes" - as in "the dust has settled". That way there is no blanket approval for changes that started simple and gained a significant change in the last commit, and leaves enough time for everybody catching the change. But then in the end the core team members have the freedom to extend the time anyway, so it might now be that relevant.

**Review by klaernie — APPROVED** [review]
generally this covers all I would like to see

File: CONTRIBUTING.md
Comment (per1234) [reply CONTRIBUTING.md:46]: I'm OK with putting it in the readme.

File: CONTRIBUTING.md
Comment (per1234) [reply CONTRIBUTING.md]: Resolved by [`923493b` (#378)](https://github.com/editorconfig-checker/editorconfig-checker/pull/378/commits/923493be7d65693b0d2cb71122ef95957da53589)

File: CONTRIBUTING.md
Comment (per1234) [reply CONTRIBUTING.md]: Resolved by [`923493b` (#378)](https://github.com/editorconfig-checker/editorconfig-checker/pull/378/commits/923493be7d65693b0d2cb71122ef95957da53589)

**Review by per1234 — APPROVED** [review]
Thanks @theoludwig!

File: CONTRIBUTING.md
Comment (ccoVeille) [inline CONTRIBUTING.md:14]:
```suggestion
All contributions are welcome from anyone willing to work in good faith with other contributors and the community. No contribution is too small, and all contributions are valued.
```

File: MAINTAINERS.md
Comment (ccoVeille) [inline MAINTAINERS.md:18]:
```suggestion
- Encourage discussions, so the best solution can emerge.
```

File: README.md
Comment (ccoVeille) [inline README.md:431]:
```suggestion
- Patch release (1.0.x -> 1.0.y)
  - Updates to format outputs (error messages, logs, ...).
  - Performance improvements which doesn't affect behavior.
  - Build process changes (e.g., updating dependencies, updating `Dockerfile`, ...).
  - Reverts (reverting a previous commit).
  - Bug fixes which result in **editorconfig-checker** reporting less linting errors, also known as "false-positives".
- Minor release (1.x.0 -> 1.y.0)
  - Adding new [configuration](./README.md#configuration) options, including new CLI flags.
  - Adding new [default path](./README.md#default-excludes) to exclude by default.
  - Adding new [format output](./README.md#output-formats).
  - Support a new [editorconfig](https://editorconfig.org/) property (e.g: `insert_final_newline`, `indent_size`, ...).
  - Any new feature which doesn't break existing behavior.
- Major release (x.0.0 -> y.0.0)
  - Removal/Rename of a [configuration](./README.md#configuration) option.
  - Removal of a [format output](./README.md#output-formats).
  - Removal of a [default path](./README.md#default-excludes) to exclude by default.
  - Removal of an existing [editorconfig](https://editorconfig.org/) property.
  - Bug fixes, which might result in **editorconfig-checker** reporting more linting errors, if the previous behavior was incorrect, with the [editorconfig](https://editorconfig.org/) specification.
```

**Review by ccoVeille — APPROVED** [review]
minor styles suggestion, otherwise LGTM :+1:

**Review by ccoVeille — APPROVED** [review]

**Review by greut — APPROVED** [review]
Thanks for this important work.

**Review by mstruebing — COMMENTED** [review]
Thanks so much for this.
Only one question, as the team is still pretty small do you think it's worth the need to have 2 approvals from maintainers?

File: MAINTAINERS.md
Comment (ccoVeille) [reply MAINTAINERS.md:9]: Honestly, rules are made to be written and read.

I don't expect to see them applied without  common sense.

So I would say this value is OK for someone reading it, so they understand the workflow and expectations.

Then PR will be merged when needed.

But maybe my visiot is this one because I'm French and you are German 😂😅

Comment (klaernie) [reply MAINTAINERS.md:9]: The strict German in me says "two are set in stone now" ;) But the other half says: the total number of team members is neither specified nor relevant - so maybe this is also a partial encouragement to create a "reviewers" role with people like @ccoVeille and me, who eagerly review code and have different views and experiences.

In the end I really would like to encourage a four-eyes principle - simple because I would not be sure that I always grasp 100% of it, no matter how much I try.

File: MAINTAINERS.md
Comment (klaernie) [inline MAINTAINERS.md:19]:
```suggestion
- Prevent issues that could arise from changes being merged too quickly.
```

File: README.md
Comment (klaernie) [inline README.md:147]: This codeblock somehow got broken - it truly is a multiline output.

File: README.md
Comment (klaernie) [inline README.md:419]: I think this makes the meaning a bit clearer
```suggestion
  - Bug fixes which result in **editorconfig-checker** reporting less linting errors (removing "false-positive" linting errors).
```

File: README.md
Comment (klaernie) [inline README.md:415]:
```suggestion
  - Updates to output formats (error messages, logs, ...).
```

File: README.md
Comment (klaernie) [inline README.md:423]:
```suggestion
  - Adding new [output formats](./README.md#output-formats).
```

File: README.md
Comment (klaernie) [inline README.md:421]:
```suggestion
  - Adding new [configuration options](./README.md#configuration).
  - Deprecation (but not their removal) of existing [configuration options](./README.md#configuration).
  - Adding new CLI flags.
```

File: README.md
Comment (klaernie) [inline README.md:427]: I think we will always follow the pattern of adding a new config option, keeping the old one valid for another release (or two), warning users that this option is deprecated and they should migrate to the new one, before we finally remove the old option. (that should maybe also be an addition to the maintainer guide.)
Hence we do not really need to mention renames, since they do not exist under this model.

```suggestion
  - Removal of a [configuration option](./README.md#configuration).
```

File: README.md
Comment (klaernie) [inline README.md:428]:
```suggestion
  - Removal of a [output format](./README.md#output-formats).
```

File: README.md
Comment (klaernie) [inline README.md:429]:
```suggestion
  - Removal of a [path to exclude by default](./README.md#default-excludes).
```

File: README.md
Comment (klaernie) [inline README.md:430]:
```suggestion
  - Removal of support for an [editorconfig](https://editorconfig.org/) property.
```

File: README.md
Comment (klaernie) [inline README.md:431]: I think I do not understand what this part means:
> if the previous behavior was incorrect, with the [editorconfig](https://editorconfig.org/) specification.

Is that "if the previous behaviour was violating the editorconfig specification"?

**Comment by klaernie** [comment]
whoops - should have put all these comments into one review. Most are just minor word-tweaking, the biggest being the question: Should we include a formal description of our deprecation cycle?

File: MAINTAINERS.md
Comment (theoludwig) [reply MAINTAINERS.md:9]: > as the team is still pretty small

Not that small. :smile: We're 9 members in the organization, and if we add @klaernie as a maintainer, we're 10 (I definitely agree to add klaernie to the organization/maintainer team).
I think, we'll need to try, and if we see it's "too much", we will be able to reduce this number, to only 1 approval.

On this PR, there are already 5 reviewers, so I feel like, it's working pretty well, and we'll be able at least to get 2 reviews per PR, and if not, because people don't have time to review "right away", it can wait a few days/weeks. I don't think we really need to urge for changes, we can wait, as editorconfig-checker is a pretty mature software already.

File: README.md
Comment (ccoVeille) [reply README.md:428]: an output format then

File: README.md
Comment (theoludwig) [reply README.md]: Oops, I've fixed it, and also fixed the "Code Climate" format, so that it has syntax highlighting fo JSON.

Comment (klaernie) [reply README.md]: Nice, wanted to suggest it, but didn't want to mix that in with this one necessarily!

File: README.md
Comment (klaernie) [reply README.md]: oh, right!

File: README.md
Comment (theoludwig) [reply README.md:430]: > Is that "if the previous behaviour was violating the editorconfig specification"?

Yes. This PR is a perfect example for that: https://github.com/editorconfig-checker/editorconfig-checker/pull/331
It's a bug fix, that resulted in **editorconfig-checker** reporting more linting errors.
This change landed in a patch version, but looking back at it, I think, it should have landed in a major version, just to be safe. I think that if **editorconfig-checker** v3.0.0 doesn't report any errors, v3.0.1 should not either.

File: MAINTAINERS.md
Comment (klaernie) [reply MAINTAINERS.md:9]: Ah, okay, that number of maintainers is surprising, given that the org only lists four (as they made their role known to the world). With that number 2 is a good number of reviews to require.

**Comment by theoludwig** [comment]
> Should we include a formal description of our deprecation cycle?

Good idea! It will also fill a bit the maintainers guidelines, as for the moment, it's pretty much empty.
Do you want to do it in a follow up PR? (as it seems like you already have ideas for that part, as you talk about it in https://github.com/editorconfig-checker/editorconfig-checker/pull/378#discussion_r1755127234.

I might do it in this PR, but maybe we can do it in a follow up PR.

File: README.md
Comment (klaernie) [inline README.md:430]: Okay, than I think this might communicate this intend better:
```suggestion
  - Bug fixes, which result in **editorconfig-checker** reporting more linting errors, even if the previous behavior was incorrect according to the [editorconfig specification](https://editorconfig.org/).
```

Comment (theoludwig) [reply README.md:430]: I'm not convinced by "even if", it's more like "because" ... the previous behavior was incorrect.

Comment (theoludwig) [reply README.md]: I've updated it to: `Bug fixes, which result in **editorconfig-checker** reporting more linting errors, because the previous behavior was incorrect according to the [editorconfig specification](https://editorconfig.org/)`.

**Comment by theoludwig** [comment]
Will merge that PR tomorrow, if that sounds good to everyone. :smile:

**Review by klaernie — APPROVED** [review]

**Review by ccoVeille — APPROVED** [review]

File: CONTRIBUTING.md
Comment (ccoVeille) [reply CONTRIBUTING.md:30]: I opened 
- https://github.com/editorconfig-checker/editorconfig-checker/issues/379

Comment (theoludwig) [reply CONTRIBUTING.md:30]: Well, it's already enforced: https://github.com/editorconfig-checker/editorconfig-checker/blob/f7b7cb7326cb657e87bc3093da4811c97892659e/.github/workflows/ci.yml#L16

Comment (theoludwig) [reply CONTRIBUTING.md:30]: https://github.com/editorconfig-checker/editorconfig-checker/pull/378#discussion_r1751349090

**Comment by klaernie** [comment]
We should set the the branch protection to reflect our decision to require two reviews - https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches#require-pull-request-reviews-before-merging
But only @editorconfig-checker/admin can do that.

**Comment by theoludwig** [comment]
> We should set the the branch protection to reflect our decision to require two reviews - https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches#require-pull-request-reviews-before-merging
> But only @editorconfig-checker/admin can do that. 

Good idea, I can't configure it, maybe @mstruebing could help here? Thanks! :smile:

**Pull request merged**
